README.md
# andbot_test_joint

## joint frame 
![](https://github.com/ChingHengWang/andbot_test_joint/blob/master/ref/joint_frame.jpeg)

## Steps
* roslaunch andbot andbot.launch or andbot_ekf.launch (ther arm arm.launch file in this two launches)

* rosrun andbot_test_joint teleop_home 0 0.1 left arm 0 Axis & 1 Axis (speed = 0.1rad/s)
  * Pres '↑' and '↓'to control the 0 Axix forward and backward, and move this Axis to 'mechanical origin point'
  * Pres '←' and '→'to control the 1 Axix left and right, and move this Axis to 'mechanical origin point'
  * press 'P' to pause the mvoement
  
* rosrun andbot_test_joint teleop_home 1 0.1 left arm 2 Axis & 3 Axis (speed = 0.1rad/s)
  * Pres '↑' and '↓'to control the 3 Axix forward and backward, and move this Axis to 'mechanical origin point'
  * Pres '←' and '→'to control the 2 Axix left and right, and move this Axis to 'mechanical origin point'
  * press 'P' to pause the mvoement
  
* rosrun andbot_test_joint teleop_home 2 0.1 right arm 0 Axis & 1 Axis (speed = 0.1rad/s)
   * Pres '↑' and '↓'to control the 0 Axix forward and backward, and move this Axis to 'mechanical origin point'
  * Pres '←' and '→'to control the 1 Axix left and right, and move this Axis to 'mechanical origin point'
  * press 'P' to pause the mvoement
  
* rosrun andbot_test_joint teleop_home 3 0.1 right arm 2 Axis & 3 Axis (speed = 0.1rad/s)
  * Pres '↑' and '↓'to control the 3 Axix forward and backward, and move this Axis to 'mechanical origin point'
  * Pres '←' and '→'to control the 2 Axix left and right, and move this Axis to 'mechanical origin point'
  * press 'P' to pause the mvoement

* Finally press 'Ctrl+C' to kill this node


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package andbot_test_joint
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.0.2 (2016-05-30)
------------------
* add arm_test.launch
* add arm.launch
* add source code(action.cpp left_arm_moveR.cpp right_arm_moveR.cpp teleop_home.cpp)
* contributors: Zach


0.0.1 (2016-05-30)
------------------
* first 


src/test.cpp
#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <signal.h>
#include <termios.h>
#include <stdio.h>
#include <sensor_msgs/Joy.h>

#define KEYCODE_R 0x43 
#define KEYCODE_L 0x44
#define KEYCODE_U 0x41
#define KEYCODE_D 0x42
#define KEYCODE_Q 0x71

#define KEYCODE_P 'p'   // for stop the robot
#define KEYCODE_A 'a'   // for starting laser scan
#define KEYCODE_S 's'   // for stopping laser scan

class TeleopHome
{
public:
  TeleopHome();
  TeleopHome(int,double);
  void keyLoop();
  int mode;
  double speed;
private:
  ros::NodeHandle nh_;
  double angular_0,angular_1;
  double angular_increment_0,angular_increment_1;
  ros::Publisher offset_pub_0,offset_pub_1;
 
};

TeleopHome::TeleopHome():
  mode(0),angular_0(0),angular_1(0),angular_increment_0(0.005),angular_increment_1(0.005)
{
if(mode==0){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/L0/cmd/offset", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/L1/cmd/offset", 1);
}
}

TeleopHome::TeleopHome(int m,double s):
  angular_0(0),angular_1(0),angular_increment_0(0.005),angular_increment_1(0.005),speed(s),mode(m)
{
if(mode==0){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/L0/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/L1/cmd/offset_v", 1);
}
if(mode==1){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/L3/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/L2/cmd/offset_v", 1);
}
if(mode==2){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/R0/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/R1/cmd/offset_v", 1);
}
if(mode==3){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/R3/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/R2/cmd/offset_v", 1);
}

}





int kfd = 0;
struct termios cooked, raw;

void quit(int sig)
{
  tcsetattr(kfd, TCSANOW, &cooked);
  ros::shutdown();
  exit(0);
}


int main(int argc, char** argv)
{
  ros::init(argc, argv, "test_sync_axis");
  int MODE=atoi(argv[1]);
  double SPEED=atof(argv[2]);
  TeleopHome teleop_home(MODE,SPEED);
  ros::Time current_time, last_time;
  current_time = ros::Time::now();
  last_time = ros::Time::now();

  signal(SIGINT,quit);

  teleop_home.keyLoop();
  
  return(0);
}


void TeleopHome::keyLoop()
{
  char c;

  // get the console in raw mode                                                              
  tcgetattr(kfd, &cooked);
  memcpy(&raw, &cooked, sizeof(struct termios));
  raw.c_lflag &=~ (ICANON | ECHO);
  // Setting a new line, then end of file                         
  raw.c_cc[VEOL] = 1;
  raw.c_cc[VEOF] = 2;
  tcsetattr(kfd, TCSANOW, &raw);

  puts("Reading from keyboard");
  puts("---------------------------");
  puts("Use arrow keys to move the joint offset.");

  ros::Rate r(50.0);

  for(;;)
  {
    // get the next event from the keyboard  
    if(read(kfd, &c, 1) < 0)
    {
      perror("read():");
      exit(-1);
    }

    ROS_DEBUG("value: 0x%02X\n", c);
  
    switch(c)
    {
      case KEYCODE_U://go
        //angular_0=angular_0+angular_increment_0;
	angular_0=speed;
	ROS_INFO("UP %f",angular_0);
        break;
      case KEYCODE_D://back
        //angular_0=angular_0-angular_increment_0;
	angular_0=-1*speed;
	ROS_INFO("DOWN %f",angular_0);
        break;

    }

    std_msgs::Float64 offset_0,offset_1;
    offset_0.data = angular_0;
    offset_1.data = angular_1;

    offset_pub_0.publish(offset_0);
    offset_pub_1.publish(offset_1);

  }


  return;
}


src/moveR.cpp
#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <geometry_msgs/Vector3.h>
#include <signal.h>
#include <termios.h>
#include <stdio.h>
#include <sensor_msgs/Joy.h>
#include <stdlib.h>
#define PI 3.14159
struct point{
  std_msgs::Float64 x;
  std_msgs::Float64 y;
  std_msgs::Float64 z;
};

struct arm_angle{
  geometry_msgs::Vector3 th_0;
  geometry_msgs::Vector3 th_1;
  geometry_msgs::Vector3 th_2;
  geometry_msgs::Vector3 th_3;
};

void v_scalar_multip(double s,double a[],double c[],int n)
{
  for (int i=0;i<n;i++)
  {
    c[i]=s*a[i];
  }

  return ;
}
double v_dot(double a[],double b[],int n)
{
  double sum=0;
  for(int i=0;i<n;i++)
  {
    sum=sum+a[i]*b[i];

   //ROS_INFO("a[i]= %f\n",a[i]);
   //ROS_INFO("b[i]= %f\n",b[i]);

  }
  return sum;
}
 
void v_sub(double a[],double b[],double c[],int n)
{
  for(int i=0;i<n;i++)
  {
    c[i]=a[i]-b[i];
  }
  return;
}
void v_add(double a[],double b[],double c[],int n)
{
  for(int i=0;i<n;i++)
  {
    c[i]=a[i]+b[i];
  }
  return;
}
void v_cross(double a[],double b[],double c[])
{
 
  c[0]=a[1]*b[2]-a[2]*b[1];
  c[1]=a[2]*b[0]-a[0]*b[2];
  c[2]=a[0]*b[1]-a[1]*b[0];

  return;
}
void v_print(double a[],int n)
{
  for(int i=0;i<n;i++)
  {
    ROS_INFO("p = %f\n",a[i]);
  }
  return;
}

class MoveR
{
public:
  point Target_left;
  arm_angle left_arm_angle;
  double lup,ldn;
  void goalCallback(const geometry_msgs::Vector3& msg);
  MoveR();
  void IK_loop();
  void IK(const point& p);
  void IK4(const point& P);

  ros::NodeHandle nh_;
  ros::Publisher axis_l0_pub,axis_l1_pub,axis_l2_pub,axis_l3_pub;
  ros::Subscriber goal_sub; 
};

 void MoveR::goalCallback(const geometry_msgs::Vector3& vector){
  //ROS_INFO("subscribe\n");
  Target_left.x.data=vector.x;
  Target_left.y.data=vector.y;
  Target_left.z.data=vector.z;
}
void MoveR::IK(const point& p)
{
  //ROS_INFO("IK\n");
  double th0=0,th1=0;
  double x=p.x.data;
  double y=p.y.data;
  double z=p.z.data;
  double temp1=pow(x,2)+pow(y,2)+pow(z,2);
  double len_E=pow(temp1,0.5);

  if(x>=0&&y<0)
    th0=atan((double)abs(x)/abs(y));
  else if(x>=0&&y>=0)
      th0=PI-atan((double)abs(x)/abs(y));
    else if(x<0&&y>=0)
        th0=PI+atan((double)abs(x)/abs(y));
      else if(x<0&&y<0)
          th0=2*PI-atan((double)abs(x)/abs(y));
  double temp2=pow(pow(x,2)+pow(y,2),0.5);
  if(z>=0)
    th1=-atan((double)abs(z)/temp2);
  else if(z<0)
    th1=atan((double)abs(z)/temp2);

   ROS_INFO("IK:th0 %f\n",th0);
   ROS_INFO("IK:th1 %f\n",th1);

  if(isnan(th0)||isnan(th1)||th0>PI||th0<-1.57||th1<0||th1>1.57){
    ROS_INFO("Fail and dont move\n");
   }
  else{
    ROS_INFO("go move\n");
    left_arm_angle.th_0.x=th0;
    left_arm_angle.th_0.y=10;
    left_arm_angle.th_1.x=th1;
    left_arm_angle.th_1.y=10;
    left_arm_angle.th_2.x=0;
    left_arm_angle.th_2.y=10;
    left_arm_angle.th_3.x=0;
    left_arm_angle.th_3.y=10;



  }
}

void MoveR::IK4(const point& P)
{
  //ROS_INFO("IK\n");
  double th0=0,th1=0,th2=0,th3=0;
  double x=P.x.data;
  double y=P.y.data;
  double z=P.z.data;
  double p[3]={x,y,z};
  double temp1=pow(x,2)+pow(y,2)+pow(z,2);
  double norm_p=pow(temp1,0.5);
  double b=acos((pow(lup,2)+pow(ldn,2)-pow(norm_p,2))/(2*lup*ldn));
  th3=PI-b;  
  double a=asin((ldn/norm_p)*sin(b));
  double s[3]={0,0,0};
  double u_z[3]={0,0,1};
  double p_s[3];
  v_sub(p,s,p_s,3);
  //ROS_INFO("p\n");
  //v_print(p,3);	
  //ROS_INFO("s\n");
  //v_print(s,3);	
  //ROS_INFO("p-s\n");
  //v_print(p_s,3);	

  double norm_p_s=pow(pow(p_s[0],2)+pow(p_s[1],2)+pow(p_s[2],2),0.5);
  double u_n[3]={p_s[0]/norm_p_s,p_s[1]/norm_p_s,p_s[2]/norm_p_s};
  
  //ROS_INFO("u_n\n");
  //v_print(u_n,3);	
  double tmp[3];
  v_scalar_multip(v_dot(u_z,u_n,3),u_n,tmp,3);
  double u[3],u_u[3];
  v_add(u_z,tmp,u,3);
  double norm_u=pow(pow(u[0],2)+pow(u[1],2)+pow(u[2],2),0.5);
  v_scalar_multip(1/norm_u,u,u_u,3);

  //ROS_INFO("u_u \n");
  //v_print(u_u,3);	
  double v[3],u_v[3];
  v_cross(u_n,u_u,v);
  double norm_v=pow(pow(v[0],2)+pow(v[1],2)+pow(v[2],2),0.5);
  v_scalar_multip(1/norm_v,v,u_v,3);

  //ROS_INFO("u_v \n");
  //v_print(u_v,3);	

  double c[3];
  v_scalar_multip(cos(a)*lup,u_n,tmp,3);
  v_add(s,tmp,c,3);
  double r=lup*sin(a);
  double fi=PI/2;
  
  //ROS_INFO("a %f\n",a);
  double tmp1[3],tmp2[3];
  v_scalar_multip(cos(fi),u_u,tmp1,3);
  v_scalar_multip(sin(fi),u_v,tmp2,3);
  v_add(tmp1,tmp2,tmp,3);
  v_scalar_multip(r,tmp,tmp,3);
  double e[3];
  v_add(c,tmp,e,3);
  //ROS_INFO("e \n");
  //v_print(e,3);	

  double ex=e[0],ey=e[1],ez=e[2];
  //ROS_INFO("ex %f\n",ex);
  //ROS_INFO("ey %f\n",ey);
  //ROS_INFO("ez %f\n",ez);

  if(ex>=0&&ey<0)
    th0=atan((double)abs(ex)/abs(ey));
  else if(ex>=0&&ey>=0)
      th0=PI-atan((double)abs(ex)/abs(ey));
    else if(ex<0&&ey>=0)
        th0=PI+atan((double)abs(ex)/abs(ey));
      else if(ex<0&&ey<0)
          th0=-atan((double)abs(ex)/abs(ey));
  double temp2=pow(pow(ex,2)+pow(ey,2),0.5);
  if(ez>=0)
    th1=-atan((double)abs(ez)/temp2);
  else if(ez<0)
    th1=atan((double)abs(ez)/temp2);

   //ROS_INFO("IK:th0 %f\n",th0);
   //ROS_INFO("IK:th1 %f\n",th1);
   //ROS_INFO("th3 %f\n",th3);


  th2=atan2(-x*sin(th0)*sin(th1)+y*cos(th0)*sin(th1)-z*cos(th1),x*cos(th0)+y*sin(th0));
  ROS_INFO("x %f  y %f  z %f\n",x,y,z);

  ROS_INFO("th0 %f  th1 %f  th2 %f  th3 %f\n",th0,th1,th2,th3);

  if(isnan(th0)||isnan(th1)||isnan(th2)||isnan(th3)||th0>PI||th0<-1.57||th1<-0.01||th1>1.57){
    ROS_INFO("Fail and dont move\n");
   }
  else{
    ROS_INFO("go move\n");
    left_arm_angle.th_0.x=th0;
    left_arm_angle.th_0.y=10;
    left_arm_angle.th_1.x=th1;
    left_arm_angle.th_1.y=10;
    left_arm_angle.th_2.x=th2;
    left_arm_angle.th_2.y=10;
    left_arm_angle.th_3.x=th3;
    left_arm_angle.th_3.y=10;

  }
}

  MoveR::MoveR()
{
  lup=235,ldn=250;    

  left_arm_angle.th_0.x=0.0;
  left_arm_angle.th_0.y=0.0;

  left_arm_angle.th_1.x=0.0;
  left_arm_angle.th_1.y=0.0;

  left_arm_angle.th_2.x=0.0;
  left_arm_angle.th_2.y=0.0;

  left_arm_angle.th_3.x=0.0;
  left_arm_angle.th_3.y=0.0;

  Target_left.x.data=0.0,Target_left.y.data=0.0,Target_left.z.data=0.0;

  axis_l0_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L0/cmd/position", 1);
  axis_l1_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L1/cmd/position", 1);
  axis_l2_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L2/cmd/position", 1);
  axis_l3_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L3/cmd/position", 1);
  goal_sub = nh_.subscribe("/andbot/left_arm/goal", 1000, &MoveR::goalCallback,this);


}


int main(int argc, char** argv)
{

  ros::init(argc, argv, "moveR_left_arm");
  MoveR move_R;
  move_R.IK_loop();
  return(0);
}


void MoveR::IK_loop()
{


  ros::Rate r(0.5);

  while(nh_.ok())
  {
    ros::spinOnce();
    IK4(Target_left);
    axis_l0_pub.publish(left_arm_angle.th_0);
    axis_l1_pub.publish(left_arm_angle.th_1);
    axis_l2_pub.publish(left_arm_angle.th_2);
    axis_l3_pub.publish(left_arm_angle.th_3);

    r.sleep();
  }


  return;
}




src/action.cpp
#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <std_msgs/UInt8.h>
#include <geometry_msgs/Vector3.h>
#include <geometry_msgs/Pose.h>
#include <stdio.h>
#include <stdlib.h>
#define PI 3.14159

unsigned int action_state=0;
unsigned int move_state=0;

struct Arm_Pose{
  double x;
  double y;
  double z;
  double fi;
  double t;
};

geometry_msgs::Pose pose_left,pose_right;
unsigned int current_time=0, last_time,duration_time=0;


Arm_Pose phome=
{0,-485,0,0,5};

struct KunFu{
  Arm_Pose left_p0,right_p0;
  Arm_Pose left_p1,right_p1;

}kunfu=
{
  {0,-350,0,0,0},{0,-350,0,0,0},
  {485,0,0,0,0},{485,0,0,0,0},
//  {200,100,0,0,3},{200,100,0,0,3},
//  {450,100,0,0,3},{450,100,0,0,3},

};

struct HoldTray{
  Arm_Pose left_p0,right_p0;
  Arm_Pose left_p1,right_p1;
}holdtray=
{
  {0,-350,0,0,0},{0,-350,0,0,0},
  {400,-200,0,0,0},{400,-200,0,0,0},
};

struct Strong{
  Arm_Pose left_p0a,right_p0a;
  Arm_Pose left_p0b,right_p0b;
  Arm_Pose left_p1a,right_p1a;
  Arm_Pose left_p1b,right_p1b;

}strong=
{
  {0,-350,0,0,0},{0,-350,0,0,0},
  {0,-350,0,0.7854,0},{0,-350,0,-0.7854,0},
  {220,180,-200,0.2854,0},{220,180,200,-0.2854,0},
  {220,180,-200,1.0,0},{220,180,200,-1.0,0},

//  {200,200,-200,2.04,0},{200,200,200,1.1,0},
};

 void predefineCallback(const std_msgs::UInt8& msg){
   action_state=msg.data;
   move_state=0;
   duration_time=0;

}


void move_to_pose_left(const Arm_Pose& pose)
{
  pose_left.position.x =pose.x; 
  pose_left.position.y =pose.y; 
  pose_left.position.z =pose.z; 
  pose_left.orientation.x =pose.fi; 
  pose_left.orientation.y =pose.t; 

}
void move_to_pose_right(const Arm_Pose& pose)
{
  pose_right.position.x =pose.x; 
  pose_right.position.y =pose.y; 
  pose_right.position.z =pose.z; 
  pose_right.orientation.x =pose.fi; 
  pose_right.orientation.y =pose.t; 
}
int main(int argc, char** argv)
{
  ros::init(argc, argv, "action");
  ros::NodeHandle nh_;
  ros::Publisher action_left_pub = nh_.advertise<geometry_msgs::Pose>("/andbot/left_arm/goal", 500);
  ros::Publisher action_right_pub = nh_.advertise<geometry_msgs::Pose>("/andbot/right_arm/goal", 500);
  ros::Subscriber pose_sub = nh_.subscribe("/andbot/predefinedPoses", 1000, predefineCallback);

  current_time = ros::Time::now().toSec();
  last_time = ros::Time::now().toSec();
  ros::Rate r(1);
  while(nh_.ok())
  {
    current_time = ros::Time::now().toSec();
    duration_time=duration_time+(current_time-last_time);
    last_time = ros::Time::now().toSec();

    ros::spinOnce();
    switch(action_state)
    {
	case 0:
            ROS_INFO("HOME(LIMP) t=%d",duration_time);
            move_to_pose_left(phome);
            move_to_pose_right(phome);
	    break;

	case 1:
 	  {
            ROS_INFO("KUN_FU t=%d",duration_time);
	    switch(move_state)
            {
              case 0:
                move_to_pose_left(kunfu.left_p0);
                move_to_pose_right(kunfu.right_p0);
                ROS_INFO("l:p0 r:p0");
                if(duration_time>=7) {move_state=1;duration_time=0;}
                break;
              case 1:
                move_to_pose_left(kunfu.left_p0);
                move_to_pose_right(kunfu.right_p1);
                ROS_INFO("l:p0 r:p1");
                if(duration_time>=15) {move_state=2;duration_time=0;}
                break;
              case 2:
                move_to_pose_left(kunfu.left_p1);
                move_to_pose_right(kunfu.right_p0);
                ROS_INFO("l:p1 r:p0");
                if(duration_time>=15) {move_state=1;duration_time=0;}
                break;


            }
	    break;

          }
	case 2:
 	  {
            ROS_INFO("HOLD_TRAY t=%d",duration_time);
	    switch(move_state)
            {

              case 0:
                move_to_pose_left(holdtray.left_p0);
                move_to_pose_right(holdtray.right_p0);
                ROS_INFO("l:p0 r:p0");
		if(duration_time>=10) {move_state=1;}
		break;
              case 1:
                move_to_pose_left(holdtray.left_p1);
                move_to_pose_right(holdtray.right_p1);
                ROS_INFO("l:p1 r:p1");
		break;

            }
	    break;

          }
	case 3:
 	  {
            ROS_INFO("Strong t=%d",duration_time);
	    switch(move_state)
            {

              case 0:
                move_to_pose_left(strong.left_p0a);
                move_to_pose_right(strong.right_p0a);
                ROS_INFO("l:p0a r:p0a");
		if(duration_time>=10) {move_state=1;duration_time=0;}
		break;
              case 1:
                move_to_pose_left(strong.left_p1a);
                move_to_pose_right(strong.right_p1a);
                ROS_INFO("l:p1a r:p1a");
		if(duration_time>=15) {move_state=0;duration_time=0;}
		break;

            }
	    break;

          }

    }

    action_left_pub.publish(pose_left);
    action_right_pub.publish(pose_right);


    r.sleep();
  }


  return(0);
}


src/Action_0.cpp
#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <geometry_msgs/Vector3.h>
#include <signal.h>
#include <termios.h>
#include <stdio.h>
#include <sensor_msgs/Joy.h>
#include <stdlib.h>
#define PI 3.14159
struct point{
  std_msgs::Float64 x;
  std_msgs::Float64 y;
  std_msgs::Float64 z;
};

struct arm_angle{
  geometry_msgs::Vector3 th_0;
  geometry_msgs::Vector3 th_1;
  geometry_msgs::Vector3 th_2;
  geometry_msgs::Vector3 th_3;
};
 
class MoveR
{
public:
  point Target_left;
  arm_angle left_arm_angle;
  double length_link_up,length_link_down;
  void goalCallback(const geometry_msgs::Vector3& msg);
  MoveR();
  void IK_loop();
  void IK(const point& p);

  ros::NodeHandle nh_;
  ros::Publisher axis_l0_pub,axis_l1_pub,axis_l2_pub,axis_l3_pub;
  ros::Subscriber goal_sub; 
};

 void MoveR::goalCallback(const geometry_msgs::Vector3& vector){
  //ROS_INFO("subscribe\n");
  Target_left.x.data=vector.x;
  Target_left.y.data=vector.y;
  Target_left.z.data=vector.z;
}
void MoveR::IK(const point& p)
{
  //ROS_INFO("IK\n");
  double th0=0,th1=0;
  double x=p.x.data;
  double y=p.y.data;
  double z=p.z.data;
  double temp1=pow(x,2)+pow(y,2)+pow(z,2);
  double len_E=pow(temp1,0.5);
  if(x>=0&&y<0)
    th0=atan((double)abs(x)/abs(y));
  else if(x>=0&&y>=0)
      th0=PI-atan((double)abs(x)/abs(y));
    else if(x<0&&y>=0)
        th0=PI+atan((double)abs(x)/abs(y));
      else if(x<0&&y<0)
          th0=2*PI-atan((double)abs(x)/abs(y));
  double temp2=pow(pow(x,2)+pow(y,2),0.5);
  if(z>=0)
    th1=-atan((double)abs(z)/temp2);
  else if(z<0)
    th1=atan((double)abs(z)/temp2);

   ROS_INFO("IK:th0 %f\n",th0);
   ROS_INFO("IK:th1 %f\n",th1);

  if(isnan(th0)||isnan(th1)||th0>PI||th0<-1.57||th1<0||th1>1.57){
    ROS_INFO("Fail and dont move\n");
   }
  else{
    ROS_INFO("go move\n");
    left_arm_angle.th_0.x=th0;
    left_arm_angle.th_0.y=10;
    left_arm_angle.th_1.x=th1;
    left_arm_angle.th_1.y=10;
    left_arm_angle.th_2.x=0;
    left_arm_angle.th_2.y=10;
    left_arm_angle.th_3.x=0;
    left_arm_angle.th_3.y=10;



  }
}
  MoveR::MoveR()
{
  length_link_up=220,length_link_down=270;    

  left_arm_angle.th_0.x=0.0;
  left_arm_angle.th_0.y=0.0;

  left_arm_angle.th_1.x=0.0;
  left_arm_angle.th_1.y=0.0;

  left_arm_angle.th_2.x=0.0;
  left_arm_angle.th_2.y=0.0;

  left_arm_angle.th_3.x=0.0;
  left_arm_angle.th_3.y=0.0;

  Target_left.x.data=0.0,Target_left.y.data=0.0,Target_left.z.data=0.0;

  axis_l0_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L0/cmd/position", 1);
  axis_l1_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L1/cmd/position", 1);
  axis_l2_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L2/cmd/position", 1);
  axis_l3_pub = nh_.advertise<geometry_msgs::Vector3>("andbot/joint/L3/cmd/position", 1);
  goal_sub = nh_.subscribe("/andbot/left_arm/goal", 1000, &MoveR::goalCallback,this);


}


int main(int argc, char** argv)
{

  ros::init(argc, argv, "moveR_left_arm");
  MoveR move_R;
//  move_R.IK_loop();
  ros::Rate r(0.5);

  while(move_R.nh_.ok())
  {
/*
    ROS_INFO("x %f\n",move_R.Target_left.x.data);i
    ROS_INFO("y %f\n",move_R.Target_left.y.data);
    ROS_INFO("z %f\n",move_R.Target_left.z.data);
*/    
    ros::spinOnce();
    move_R.IK(move_R.Target_left);
    move_R.axis_l0_pub.publish(move_R.left_arm_angle.th_0);
    move_R.axis_l1_pub.publish(move_R.left_arm_angle.th_1);
    move_R.axis_l2_pub.publish(move_R.left_arm_angle.th_2);
    move_R.axis_l3_pub.publish(move_R.left_arm_angle.th_3);

    r.sleep();
  }
  return(0);
}


void MoveR::IK_loop()
{


  ros::Rate r(50.0);

  while(nh_.ok())
  {
    ros::spinOnce; 
    IK(Target_left);
    axis_l0_pub.publish(Target_left.x);
    axis_l1_pub.publish(left_arm_angle.th_1);
    axis_l2_pub.publish(left_arm_angle.th_2);
    axis_l3_pub.publish(left_arm_angle.th_3);

    r.sleep();
  }


  return;
}




src/teleop_home.cpp
#include <ros/ros.h>ngular_increment
#include <std_msgs/Float64.h>
#include <signal.h>
#include <termios.h>
#include <stdio.h>
#include <sensor_msgs/Joy.h>

#define KEYCODE_R 0x43 
#define KEYCODE_L 0x44
#define KEYCODE_U 0x41
#define KEYCODE_D 0x42
#define KEYCODE_Q 0x71

#define KEYCODE_P 'p'   // for stop the robot
#define KEYCODE_A 'a'   // for starting laser scan
#define KEYCODE_S 's'   // for stopping laser scan

class TeleopHome
{
public:
  TeleopHome();
  TeleopHome(int,double);
  void keyLoop();
  int mode;
  double speed;
private:
  ros::NodeHandle nh_;
  double angular_0,angular_1;
  double angular_increment_0,angular_increment_1;
  ros::Publisher offset_pub_0,offset_pub_1;
 
};

TeleopHome::TeleopHome():
  mode(0),angular_0(0),angular_1(0),angular_increment_0(0.005),angular_increment_1(0.005)
{
if(mode==0){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/L0/cmd/offset", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/L1/cmd/offset", 1);
}
}

TeleopHome::TeleopHome(int m,double s):
  angular_0(0),angular_1(0),angular_increment_0(0.005),angular_increment_1(0.005),speed(s),mode(m)
{
if(mode==0){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/L0/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/L1/cmd/offset_v", 1);
}
if(mode==1){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/L3/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/L2/cmd/offset_v", 1);
}
if(mode==2){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/R0/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/R1/cmd/offset_v", 1);
}
if(mode==3){
  offset_pub_0 = nh_.advertise<std_msgs::Float64>("andbot/joint/R3/cmd/offset_v", 1);
  offset_pub_1 = nh_.advertise<std_msgs::Float64>("andbot/joint/R2/cmd/offset_v", 1);
}

}





int kfd = 0;
struct termios cooked, raw;

void quit(int sig)
{
  tcsetattr(kfd, TCSANOW, &cooked);
  ros::shutdown();
  exit(0);
}


int main(int argc, char** argv)
{
  ros::init(argc, argv, "teleop_home");
  int MODE=atoi(argv[1]);
  double SPEED=atof(argv[2]);
  TeleopHome teleop_home(MODE,SPEED);
  ros::Time current_time, last_time;
  current_time = ros::Time::now();
  last_time = ros::Time::now();

  signal(SIGINT,quit);

  teleop_home.keyLoop();
  
  return(0);
}


void TeleopHome::keyLoop()
{
  char c;

  // get the console in raw mode                                                              
  tcgetattr(kfd, &cooked);
  memcpy(&raw, &cooked, sizeof(struct termios));
  raw.c_lflag &=~ (ICANON | ECHO);
  // Setting a new line, then end of file                         
  raw.c_cc[VEOL] = 1;
  raw.c_cc[VEOF] = 2;
  tcsetattr(kfd, TCSANOW, &raw);

  puts("Reading from keyboard");
  puts("---------------------------");
  puts("Use arrow keys to move the joint offset.");

  ros::Rate r(50.0);

  for(;;)
  {
    // get the next event from the keyboard  
    if(read(kfd, &c, 1) < 0)
    {
      perror("read():");
      exit(-1);
    }

    ROS_DEBUG("value: 0x%02X\n", c);
  
    switch(c)
    {
      case KEYCODE_U:
        //angular_0=angular_0+angular_increment_0;
	angular_0=speed;
	ROS_INFO("UP %f",angular_0);
        break;
      case KEYCODE_D:
        //angular_0=angular_0-angular_increment_0;
	angular_0=-1*speed;
	ROS_INFO("DOWN %f",angular_0);
        break;
      case KEYCODE_R:
        //angular_1=angular_1-angular_increment_1;
	angular_1=-1*speed;
	ROS_INFO("RIGHT %f",angular_1);
        break;
      case KEYCODE_L:
        //angular_1=angular_1+angular_increment_1;
	angular_1=speed;
	ROS_INFO("LEFT %f",angular_1);
        break;
      case KEYCODE_P:
        ROS_INFO("STOP MOVING");
        angular_0 = 0.0;
	angular_1 = 0.0;
        break;


    }

    std_msgs::Float64 offset_0,offset_1;
    offset_0.data = angular_0;
    offset_1.data = angular_1;

    offset_pub_0.publish(offset_0);
    offset_pub_1.publish(offset_1);

  }


  return;
}
